orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -706,13 +827,39 @@ Status ToCheckpointState(gsl::span<const uint8_t> checkpoint_bytes, CheckpointSt
                     "Checkpoint is invalid. Checkpoint version ", checkpoint_version, " is not supported.");
 
   const auto* fbs_module_state = fbs_checkpoint->module_state();
+
+  fbs::utils::ExternalDataReader external_data_reader = nullptr;
+  std::optional<std::ifstream> external_data_stream;
+
+  if (fbs_module_state->has_external_data()) {
+    ORT_RETURN_IF_NOT(checkpoint_path.has_value(),
+                      "External data is present in the checkpoint but the checkpoint path is not provided. External data with loading from buffer is not supported yet.");
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path.value());
+    external_data_stream = std::ifstream(data_path, std::ios::binary);
+
+    char errbuf[256];
+    ORT_RETURN_IF(external_data_stream.value().fail(),
+                  "Failed to open checkpoint's external data file: ", ToUTF8String(data_path), strerror_s(errbuf, sizeof(errbuf), errno));
+
+    external_data_reader = [&external_data_stream](uint64_t offset, gsl::span<uint8_t> output_buffer) {
```

Comment:
Since we have defined the same lambda twice now in this file, shall we create a single function somewhere and move all the common code there?

Reply:
I believe that equates to using a lambda for the capture, and having a static local function that does the work including error checking.

i.e.
```c++
    external_data_reader = [&external_data_stream](uint64_t offset, gsl::span<uint8_t> output_buffer) {
      return read_from_external_file_helper(external_data_stream, offset, output_buffer);
    }
```